File: Day #654/Solution.java
// Smallest window containing every distinct char: sliding window with two pointers.
// Expand right until all distinct chars present, shrink left to minimize. Time O(n), space O(k).
import java.util.*;

public class Solution {
    static int smallestWindow(String s) {
        Set<Character> distinct = new HashSet<>();
        for (char c : s.toCharArray()) distinct.add(c);
        int need = distinct.size();
        Map<Character, Integer> cnt = new HashMap<>();
        int have = 0, best = Integer.MAX_VALUE, left = 0;
        for (int right = 0; right < s.length(); right++) {
            char c = s.charAt(right);
            cnt.merge(c, 1, Integer::sum);
            if (cnt.get(c) == 1) have++;
            while (have == need) {
                best = Math.min(best, right - left + 1);
                char lc = s.charAt(left);
                cnt.merge(lc, -1, Integer::sum);
                if (cnt.get(lc) == 0) have--;
                left++;
            }
        }
        return best == Integer.MAX_VALUE ? 0 : best;
    }

    public static void main(String[] args) {
        System.out.println(smallestWindow("jiujitsu"));
    }
}

File: Day #654/solution.js
// Smallest window containing every distinct char: sliding window with two pointers.
// Expand right until all distinct chars present, shrink left to minimize. Time O(n), space O(k).
function smallestWindow(s) {
  const need = new Set(s).size;
  const cnt = new Map();
  let have = 0, best = Infinity, left = 0;
  for (let right = 0; right < s.length; right++) {
    const c = s[right];
    cnt.set(c, (cnt.get(c) || 0) + 1);
    if (cnt.get(c) === 1) have++;
    while (have === need) {
      best = Math.min(best, right - left + 1);
      const lc = s[left];
      cnt.set(lc, cnt.get(lc) - 1);
      if (cnt.get(lc) === 0) have--;
      left++;
    }
  }
  return best === Infinity ? 0 : best;
}

console.log(smallestWindow("jiujitsu"));
